design-proposals/klip-9-table-functions.md

```diff
@@ -0,0 +1,347 @@
+ which pulls out the struct fields.
+
+## Value/Return
+
+This functionality will make it easier for users to make end-end streaming apps using KSQL
+alone without having to have custom stages using Kafka Streams.
+
+We believe its a common case for users to have array or map like data structures as values
+in messages that they wish to explode. This new feature would allow them to do that in a pure KSQL
+app thus allowing them to create richer KSQL applications.
+
+## Public APIs
+
+There will be no changes to the KSQL language or any existing public APIs or configuration files.
+
+We will introduce new annotations to allow users to mark their UDTFs, in a very similar way to
+UDFs.
+
+## Design
+
+### General plumbing
+
+We will need the following general plumbing to support table functions:
+
+* In a similar way to how UDFs are currently supported we will need to create a 
+`KsqlTableFunction` which is a wrapper around the actual table function (whether built-in or
+generated). We'll also need to augment the function registry to manage the addition and retrieval
+of table functions.
+* Augment the query analysis code to extract information relating to any table functions
+in the query. We can do this either by extending the existing query analyzer or perhaps creating
+a new `TableFunctionAnalyzer` in a similar way to how aggregate analysis uses its own 
+`AggregateAnalyzer`.
+* Augment the building of logical plan by building a new node representing a FlatMap stage of the
+stream processing in the case there are table functions in the SELECT.
+* Augment the execution plan logic by adding new nodes in the execution plan corresponding to any
+extra FlatMap stage required by the query.
+* Amend the columns used in the final projection to reflect any generated columns created by a 
+FlatMap stage.
+* Implement the actual FlatMap logic. For each row, any table functions on that row will be evaluated.
+The number of rows returned from the FlatMap will be equal to the greatest of the number of rows
+returned from any of the table functions for that row. Any non table function values for the row will
+be duplicated for each row. For any table function that returned fewer than the greatest number of rows
+the values for that columnn will be `null`. For performance reasons, the FlatMap logic should cache any
+ table functions for the columns to avoid them being looked up each time.
+ 
+### UDTF API
+
+Users will be able to implement their own table function by creating their classes containing
+methods annotated with `@Udtf`.
+
+A UDTF method can contain zero or more parameters, of any of the valid KSQL types mapped to Java.
+The return value must be a `List<T>` where T is any valid KSQL type mapped to Java. 
+
+```
+@UdtfDescription(name = "SplitSpring", author = "bob", description = "Splits a string into words")
+public class SplitString {
+
+    @Udtf(description="Splits a string into words")
+    public List<String> split(String input) {
```

Comment:
probably goes without saying that we should make sure we support generics, structs and `@UdfSchemaProvider` in these APIs alongside `@UdtfParameter` or something like that  for the parameter documentation (and that all of these help contribute to the DESCRIBE FUNCTION call)